operator/podinfo/podinfo_controller.go

```diff
@@ -0,0 +1,143 @@
+	"github.com/google/go-cmp/cmp"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// Reconciler reconciles a PodInfo object
+type Reconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=cilium.io,resources=PodInfo,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=cilium.io,resources=PodInfo/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=cilium.io,resources=PodInfo/finalizers,verbs=update
+
+// Reconcile gets notified about a pod and reconciles the corresponding PodInfo object.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx)
+
+	// Get the Pod.
+	pod := &corev1.Pod{}
+	if err := r.Get(ctx, req.NamespacedName, pod); err != nil {
+		if !errors.IsNotFound(err) {
+			// Error fetching the pod. Try again later.
+			l.Error(err, "unable to fetch Pod")
+			return ctrl.Result{}, err
+		}
+		// Pod is deleted. Nothing to reconcile.
+		return ctrl.Result{}, nil
+	}
+
+	// Wait until the necessary pod fields are available.
+	if !hasAllRequiredFields(pod) {
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	podInfo := &ciliumiov1alpha1.PodInfo{}
+	if err := r.Get(ctx, req.NamespacedName, podInfo); err != nil {
+		if !errors.IsNotFound(err) {
+			// Error fetching the pod info. Try again later.
+			return ctrl.Result{}, err
+		}
+		// Pod info does not exist. Create it.
+		return ctrl.Result{}, r.Create(ctx, generatePodInfo(pod))
+	}
+	if !equal(pod, podInfo) {
+		updatedPodInfo := generatePodInfo(pod)
+		updatedPodInfo.ResourceVersion = podInfo.ResourceVersion
+		return ctrl.Result{}, r.Update(ctx, updatedPodInfo)
+	}
+	return ctrl.Result{}, nil
+}
+
+// equal returns true if the given pod and pod info are equal.
+func equal(pod *corev1.Pod, podInfo *ciliumiov1alpha1.PodInfo) bool {
+	if len(pod.Status.PodIPs) != len(podInfo.Status.PodIPs) {
+		return false
+	}
+	for i, podIP := range pod.Status.PodIPs {
+		if podIP.IP != podInfo.Status.PodIPs[i].IP {
+			return false
+		}
+	}
+	return pod.Name == podInfo.Name &&
+		pod.Namespace == podInfo.Namespace &&
+		pod.Status.PodIP == podInfo.Status.PodIP &&
+		cmp.Equal(pod.Annotations, podInfo.Annotations) &&
+		cmp.Equal(pod.Labels, podInfo.Labels) &&
+		len(podInfo.OwnerReferences) == 1 &&
+		podInfo.OwnerReferences[0].Name == pod.Name &&
+		podInfo.OwnerReferences[0].UID == pod.UID &&
+		podInfo.OwnerReferences[0].Kind == "Pod" &&
+		podInfo.OwnerReferences[0].APIVersion == "v1" &&
+		podInfo.OwnerReferences[0].Controller != nil &&
+		*podInfo.OwnerReferences[0].Controller &&
+		podInfo.OwnerReferences[0].BlockOwnerDeletion != nil &&
+		*podInfo.OwnerReferences[0].BlockOwnerDeletion
```

Comment:
Can we use reflect.DeepEqual on `OwnerReferences`? Or, if it's problematic, extract this part to a separate function maybe. I got confused for a moment why all these checks are here.

Reply:
I am not sure if we can use DeepEqual here because we are comparing two separate things (Pod and PodInfo). I can separate out the logic for ownerReference.

Reply:
I see. Yeah, `reflect.DeepEqual` will always return false if we compare Pod and PodInfo, but we can use it to compare only OwnerReferences, which are same type in both I believe.

Reply:
sure, let me try both and see what works. I am a bit unsure because `podInfo` are owned by `Pods`, not the other way around, so their OwnerReferences might not match.

Reply:
Ah, sorry, I made it confusing. Yes, OwnerReferences will be very different for Pod and PodInfo. We should not compare these two, but compare PodInfo OwnerReferenes to an OwnerReferences literal (as you did in this PR, just I would use a different syntax for that).

Reply:
Not a problem ! So, will a separate function work ?

Reply:
i think what anna is suggesting is to define an expected owners reference like this:
```
	controller := true
	blockOwnerDeletion := true
	expectedOwnerReference := metav1.OwnerReference{
		APIVersion:         "v1",
		Kind:               "Pod",
		Name:               pod.Name,
		UID:                pod.UID,
		Controller:         &controller,
		BlockOwnerDeletion: &blockOwnerDeletion,
	}
```
and then compare it against podinfo owner reference using `reflect.DeepEqual`.